### ISSUE TITLE
zsysctl boot commit fails, no new grub menu entries, 

### ISSUE THREAD

**Issue by averyfreeman** [ubuntu/zsys#168]
Hi,

I just recovered from a non-booting system. It was a default installation of `ubuntu 20.10 zfs` that had been working fine for a couple days, but I'm fairly certain what caused it to fail booting is `update-grub` is not creating new menu entries, and `apt autoremove` deleted initrd file `grub.cfg` was pointing to.

`grub-mkconfig` appears not to be able to look at `bpool`.  I'm really confused about what's going on, but it seems to be an ongoing issue, and I believe it originated with `zsys`.  I hope someone can help me.

```
# update-grub
Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/init-select.cfg'
Generating grub configuration file ...
Some pools couldn't be imported and will be ignored:
cannot import 'bpool': pool already exists    # I think this is the main issue (?)
cannot import 'rpool': pool already exists
Found linux image: vmlinuz-5.8.0-23-generic in rpool/ROOT/ubuntu_84vh5h
Found initrd image: initrd.img-5.8.0-23-generic in rpool/ROOT/ubuntu_84vh5h
Found linux image: vmlinuz-5.8.0-22-generic in rpool/ROOT/ubuntu_84vh5h
Found initrd image: initrd.img-5.8.0-22-generic in rpool/ROOT/ubuntu_84vh5h
Found linux image: vmlinuz-5.8.0-23-generic in rpool/ROOT/ubuntu_84vh5h@autozsys_6llbir

(truncated)
```

There's a lot of moving parts, but here's the reason I'm wondering if it has to do with `zsys` was because `zsysctl boot commit` failed, and `zsysctl machine show` does not show `bpool`

Edit: I should clarify, `zsysctl boot commit` failed earlier, but I no longer have the scroll buffer, sorry I cannot share the output with you.  Now it says `zsys-meta:modified-boot`, which appears somewhat normal. 

```
# zsysctl -vv machine show 
DEBUG /zsys.Zsys/MachineShow() call logged as [c0e2688b:46dd7a84] 
DEBUG Check if grpc request peer is authorized     
DEBUG Authorized as being administrator            
INFO Retrieving information for machine rpool/ROOT/ubuntu_84vh5h 
Name:           rpool/ROOT/ubuntu_84vh5h
ZSys:           true
Last Used:      current
History:        
  - Name:       rpool/ROOT/ubuntu_84vh5h@autozsys_ay9l5x
    Created on: 2020-10-15 13:14:07
  - Name:       rpool/ROOT/ubuntu_84vh5h@autozsys_7j9b5k
    Created on: 2020-10-15 12:45:53
  - Name:       rpool/ROOT/ubuntu_84vh5h@autozsys_m2ibfv
    Created on: 2020-10-15 12:35:39
  - Name:       rpool/ROOT/ubuntu_84vh5h@autozsys_6llbir
    Created on: 2020-10-15 12:07:41
Users:
  - Name:    avery
    History: 
     - rpool/USERDATA/avery_5merai@autozsys_ay9l5x (2020-10-15 13:14:10)
     - rpool/USERDATA/avery_5merai@autozsys_8gyrt8 (2020-10-15 13:08:26)
     - rpool/USERDATA/avery_5merai@autozsys_7j9b5k (2020-10-15 12:45:57)
     - rpool/USERDATA/avery_5merai@autozsys_m2ibfv (2020-10-15 12:35:42)
     - rpool/USERDATA/avery_5merai@autozsys_9qrp3j (2020-10-15 12:07:59)
     - rpool/USERDATA/avery_5merai@autozsys_6llbir (2020-10-15 12:07:59)
  - Name:    root
    History: 
     - rpool/USERDATA/root_5merai@autozsys_ay9l5x (2020-10-15 13:14:11)
     - rpool/USERDATA/root_5merai@autozsys_7j9b5k (2020-10-15 12:45:57)
     - rpool/USERDATA/root_5merai@autozsys_m2ibfv (2020-10-15 12:35:42)
     - rpool/USERDATA/root_5merai@autozsys_6llbir (2020-10-15 12:07:59)
```

I also noticed this:

```
# zsysctl service -vv reload
DEBUG /zsys.Zsys/Reload() call logged as [53cc97a1:18b77189] 
DEBUG Check if grpc request peer is authorized     
DEBUG Authorized as being administrator            
INFO Reloading daemon configuration               
DEBUG couldn't find default configuration path, fallback to internal default
```

I
A couple things I am wondering about: 

should `zsys` be seeing my `bpool`?

should zsys have a default configuration path?

There's no zfs inherit `bpool/BOOT/ubuntu_84vh5h/grub` dataset.  From this reference:  https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html

They point to  `com.ubuntu.zsys:bootfs bpool/BOOT/ubuntu_UUID/grub`.  Is this a dataset that was destroyed somehow?

I'm attaching a few captured scroll buffers as txt files since they're so large - I also took notes throughout the recovery process (from boot USB) I'm including just in case they're helpful

[env.txt](https://github.com/ubuntu/zsys/files/5387745/env.txt)
[zsysctl-boot-commit-fails.txt](https://github.com/ubuntu/zsys/files/5387753/zsysctl-boot-commit-fails.txt)
[grub-try-fix-process.txt](https://github.com/ubuntu/zsys/files/5387805/grub-try-fix-process.txt)

**Comment by ddnexus** [comment]
> should zsys be seeing my bpool?

zsys sees and uses the bpool. If you have a default installation you should have no problems with it.

> There's no zfs inherit bpool/BOOT/ubuntu_84vh5h/grub dataset. From this reference: https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html
> They point to com.ubuntu.zsys:bootfs bpool/BOOT/ubuntu_UUID/grub. Is this a dataset that was destroyed somehow?

That should only apply to the suggested mirror setup, which even for mirror is wrong and does not work, so if you don't have mirror, and you don't have that dataset it's all right!

**Comment by didrocks** [comment]
Can you please copy here the content as described in the bug template so that we can see what layout you have:
**For ubuntu users, please run and  copy the following:**
1. `ubuntu-bug zsys --save=/tmp/report`
2. Copy paste below `/tmp/report` content:
```
COPY REPORT CONTENT HERE.
```

This will help seeing what’s happening. You say you have a default installation of ubuntu 20.10 zfs, was it setup with ubiquity and the ZFS option? Indeed, it seems that /boot isn’t mounted from bpool for you and that’s puzzling, this is probably what your issue is.
You mention a "bpool/BOOT/ubuntu_84vh5h/grub" which isn’t part of the standard, default installation made by ubiquity, I wonder if you changed anything manually that could have broken it?

**Comment by averyfreeman** [comment]
> >  *OpenZFS Ubuntu ZoL Guide* points to com.ubuntu.zsys:bootfs bpool/BOOT/ubuntu_UUID/grub. 
>
> That should only apply to the suggested mirror setup, which even for mirror is wrong and does not work, so if you don't have mirror, and you don't have that dataset it's all right!

Nope, I don't have a mirror, just a single NVMe drive.  Thanks for clearing that up.

**Comment by averyfreeman** [comment]
> You say you have a default installation of ubuntu 20.10 zfs, was it setup with ubiquity and the ZFS option? Indeed, it seems that /boot isn’t mounted from bpool for you and that’s puzzling, this is probably what your issue is.
> You mention a "bpool/BOOT/ubuntu_84vh5h/grub" which isn’t part of the standard, default installation made by ubiquity, I wonder if you changed anything manually that could have broken it?

Hi, no I haven't changed anything manually.  Yes, it was the Ubuntu installer (I wasn't aware it had a name).  The ISO was downloaded from the current daily builds repo: http://cdimage.ubuntu.com/daily-live/current/

Ubuntu installer created an EFI partition, but referred to the EFI on another disk.  I had to reconfigure it to point to the same drive Ubuntu was installed on when I did the recovery.  I installed Grub on the Ubuntu drive's EFI, and also changed the `/etc/fstab` to mount that drive's EFI, because it had been mounting the other drive's EFI part.

I did that during a fairly standard recovery process:
 - boot Ubuntu ISO from USB
 - import rpool to /mnt
 - import bpool to /mnt/boot
 - import EFI to /mnt/boot/efi
 - mount /proc /sys /dev to /mnt
 - chroot to /mnt and run grub installer
 - installed systemd-boot as a backup

It was more involved than that obviously, but that's the basic synopsis.  I'm still using systemd-boot since Grub is still not updating (and thus, the kernel it's referring to has been removed by `apt autoremove` again).  I'm really glad I did!

I threw an EXT4 copy of Ubuntu on the other drive in my laptop so I can boot it in case the ZFS drive won't boot again... :/

What am I looking for in here?  I have to trim it down since it's too long for a comment, I'm going to try nuking the snapshots, hopefully it'll fit:

```
coming soon
```

**Comment by didrocks** [comment]
So, looking at your installation:
- there is no `bpool/BOOT/ubuntu_.../grub` dataset (which is what I expect from our default installation and is right), you mentioned it above but I guess you nuked it, this isn’t the source of the problem though
- you have the grub menu and firmware code on `/dev/nvme1n1p1`, which is the EFI partition as well. Again, this is what we expect by default.
- your `grub.cfg` there is correct (`/boot/grub/grub.cfg`) and contains many entries (normal, advanced and 4 history):
```
menuentry 'Ubuntu 20.10' --class ubuntu --class gnu-linux --class gnu --class os ${menuentry_id_option} 'gnulinux-rpool/ROOT/ubuntu_84vh5h-5.8.0-25-generic' {
 	recordfail
 	load_video
 	gfxmode ${linux_gfx_mode}
 	insmod gzio
 	if [ "${grub_platform}" = xen ]; then insmod xzio; insmod lzopio; fi
 	insmod part_gpt
 	insmod zfs
 	if [ x$feature_platform_search_hint = xy ]; then
 	  search --no-floppy --fs-uuid --set=root  c25c5e527df3275b
 	else
 	  search --no-floppy --fs-uuid --set=root c25c5e527df3275b
 	fi
 	linux	/BOOT/ubuntu_84vh5h@/vmlinuz-5.8.0-25-generic root=ZFS=rpool/ROOT/ubuntu_84vh5h ro   
 	initrd	/BOOT/ubuntu_84vh5h@/initrd.img-5.8.0-25-generic
 }
 submenu 'Advanced options for Ubuntu 20.10' ${menuentry_id_option} 'gnulinux-advanced-rpool/ROOT/ubuntu_84vh5h' {
 	menuentry 'Ubuntu 20.10, with Linux 5.8.0-25-generic' --class ubuntu --class gnu-linux --class gnu --class os ${menuentry_id_option} 'gnulinux-rpool/ROOT/ubuntu_84vh5h-5.8.0-25-generic' {
 		recordfail
 		load_video
 		gfxmode ${linux_gfx_mode}
 		insmod gzio
 		if [ "${grub_platform}" = xen ]; then insmod xzio; insmod lzopio; fi
 		insmod part_gpt
 		insmod zfs
 		if [ x$feature_platform_search_hint = xy ]; then
 		  search --no-floppy --fs-uuid --set=root  c25c5e527df3275b
 		else
 		  search --no-floppy --fs-uuid --set=root c25c5e527df3275b
 		fi
 		echo Loading Linux 5.8.0-25-generic ...
 		linux	/BOOT/ubuntu_84vh5h@/vmlinuz-5.8.0-25-generic root=ZFS=rpool/ROOT/ubuntu_84vh5h ro   
 		echo 'Loading initial ramdisk ...'
 		initrd	/BOOT/ubuntu_84vh5h@/initrd.img-5.8.0-25-generic
 	}
 	menuentry 'Ubuntu 20.10, with Linux 5.8.0-25-generic (recovery mode)' --class ubuntu --class gnu-linux --class gnu --class os ${menuentry_id_option} 'gnulinux-rpool/ROOT/ubuntu_84vh5h-5.8.0-25-generic' {
 		recordfail
 		load_video
 		insmod gzio
 		if [ "${grub_platform}" = xen ]; then insmod xzio; insmod lzopio; fi
 		insmod part_gpt
 		insmod zfs
 		if [ x$feature_platform_search_hint = xy ]; then
 		  search --no-floppy --fs-uuid --set=root  c25c5e527df3275b
 		else
 		  search --no-floppy --fs-uuid --set=root c25c5e527df3275b
 		fi
 		echo Loading Linux 5.8.0-25-generic ...
 		linux	/BOOT/ubuntu_84vh5h@/vmlinuz-5.8.0-25-generic root=ZFS=rpool/ROOT/ubuntu_84vh5h ro recovery nomodeset dis_ucode_ldr  
 		echo 'Loading initial ramdisk ...'
 		initrd	/BOOT/ubuntu_84vh5h@/initrd.img-5.8.0-25-generic
 	}
 	menuentry 'Ubuntu 20.10, with Linux 5.8.0-23-generic' --class ubuntu --class gnu-linux --class gnu --class os ${menuentry_id_option} 'gnulinux-rpool/ROOT/ubuntu_84vh5h-5.8.0-23-generic' {
 		recordfail
 		load_video
 		gfxmode ${linux_gfx_mode}
 		insmod gzio
 		if [ "${grub_platform}" = xen ]; then insmod xzio; insmod lzopio; fi
 		insmod part_gpt
 		insmod zfs
 		if [ x$feature_platform_search_hint = xy ]; then
 		  search --no-floppy --fs-uuid --set=root  c25c5e527df3275b
 		else
 		  search --no-floppy --fs-uuid --set=root c25c5e527df3275b
 		fi
 		echo Loading Linux 5.8.0-23-generic ...
 		linux	/BOOT/ubuntu_84vh5h@/vmlinuz-5.8.0-23-generic root=ZFS=rpool/ROOT/ubuntu_84vh5h ro   
 		echo 'Loading initial ramdisk ...'
 		initrd	/BOOT/ubuntu_84vh5h@/initrd.img-5.8.0-23-generic
 	}
 	menuentry 'Ubuntu 20.10, with Linux 5.8.0-23-generic (recovery mode)' --class ubuntu --class gnu-linux --class gnu --class os ${menuentry_id_option} 'gnulinux-rpool/ROOT/ubuntu_84vh5h-5.8.0-23-generic' {
 		recordfail
 		load_video
 		insmod gzio
 		if [ "${grub_platform}" = xen ]; then insmod xzio; insmod lzopio; fi
 		insmod part_gpt
 		insmod zfs
 		if [ x$feature_platform_search_hint = xy ]; then
 		  search --no-floppy --fs-uuid --set=root  c25c5e527df3275b
 		else
 		  search --no-floppy --fs-uuid --set=root c25c5e527df3275b
 		fi
 		echo Loading Linux 5.8.0-23-generic ...
 		linux	/BOOT/ubuntu_84vh5h@/vmlinuz-5.8.0-23-generic root=ZFS=rpool/ROOT/ubuntu_84vh5h ro recovery nomodeset dis_ucode_ldr  
 		echo 'Loading initial ramdisk ...'
 		initrd	/BOOT/ubuntu_84vh5h@/initrd.img-5.8.0-23-generic
 	}
 }
 submenu 'History for Ubuntu 20.10' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_84vh5h' {
 	submenu 'Revert to 10/20/2020 @ 02:01' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_84vh5h@autozsys_qhozo5' {
 		zsyshistorymenu "rpool/ROOT/ubuntu_84vh5h@autozsys_qhozo5" "/dev/nvme1n1p3" "/BOOT/ubuntu_84vh5h@autozsys_qhozo5/initrd.img-5.8.0-25-generic" "/BOOT/ubuntu_84vh5h@autozsys_qhozo5/vmlinuz-5.8.0-25-generic" "5.8.0-25-generic"
 	}
 	submenu 'Revert to 10/20/2020 @ 01:56' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_84vh5h@autozsys_5rzs51' {
 		zsyshistorymenu "rpool/ROOT/ubuntu_84vh5h@autozsys_5rzs51" "/dev/nvme1n1p3" "/BOOT/ubuntu_84vh5h@autozsys_5rzs51/initrd.img-5.8.0-25-generic" "/BOOT/ubuntu_84vh5h@autozsys_5rzs51/vmlinuz-5.8.0-25-generic" "5.8.0-25-generic"
 	}
 	submenu 'Revert to 10/19/2020 @ 22:36' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_84vh5h@autozsys_iizkei' {
 		zsyshistorymenu "rpool/ROOT/ubuntu_84vh5h@autozsys_iizkei" "/dev/nvme1n1p3" "/BOOT/ubuntu_84vh5h@autozsys_iizkei/initrd.img-5.8.0-25-generic" "/BOOT/ubuntu_84vh5h@autozsys_iizkei/vmlinuz-5.8.0-25-generic" "5.8.0-25-generic"
 	}
 	submenu 'Revert to 10/19/2020 @ 22:23' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_84vh5h@autozsys_1rwx3i' {
 		zsyshistorymenu "rpool/ROOT/ubuntu_84vh5h@autozsys_1rwx3i" "/dev/nvme1n1p3" "/BOOT/ubuntu_84vh5h@autozsys_1rwx3i/initrd.img-5.8.0-25-generic" "/BOOT/ubuntu_84vh5h@autozsys_1rwx3i/vmlinuz-5.8.0-25-generic" "5.8.0-25-generic"
 	}
 }
```
The upgrade-grub warning is thus weird, as you have this grub.cfg generated. It seems (the os-prober part) that you have a secondary installation detected, unsure why. However, all that doesn’t explain why you don’t see your grub.cfg content.
My bet is that your system grub is not pointing to the correct one (basically, you have 2 grub install) and you should try to run `grub-install` with the correct partition so that your efit and gpt points to the correct grub installation.

**Comment by averyfreeman** [comment]
>     * there is no `bpool/BOOT/ubuntu_.../grub` dataset (which is what I expect from our default installation and is right), you mentioned it above but I guess you nuked it, this isn’t the source of the problem though

Just to be clear, I never nuked any `bpool/*` datasets explicitly. I'm still confused about whether `bpool/BOOT/ubuntu_.../grub` dataset is installed by default, I thought you said it wasn't installed by default (?) I just saw it in an OpenZFS guide for installing ZFS manually on Ubuntu.

`os-prober` could possibly be seeing the Ubuntu EXT4 installation on `/dev/nvme0n1` (which has its own UEFI partition), or perhaps the `systemd-boot`, UEFI bootloaders on `/dev/nvme1n1p1`.

Will try and run `grub-install` again and report back.  I guess this isn't an issue w/ `zsys` after all, so I imagine we should probably close this, but I greatly appreciate your help.   I'll wait to close it myself until I report back with `grub-install` results, but if you want to close it before that I completely understand.

**Comment by averyfreeman** [comment]
>     * there is no `bpool/BOOT/ubuntu_.../grub` dataset (which is what I expect from our default installation and is right), you mentioned it above but I guess you nuked it, this isn’t the source of the problem though

Just to be clear, I never nuked any `bpool/*` datasets explicitly. I'm still confused about whether `bpool/BOOT/ubuntu_.../grub` dataset is installed by default, I thought you said it wasn't installed by default (?) I just saw it in an OpenZFS guide for installing ZFS manually on Ubuntu.

`os-prober` could have possibly been seeing the Ubuntu EXT4 installation on `/dev/nvme0n1` (which has its own UEFI partition), or perhaps the `systemd-boot`, UEFI bootloaders on `/dev/nvme1n1p1`.

I ran `grub-install` again and it appears to have worked.  `os-prober` found my other ubuntu installation on another drive (EXT4) and I have all the snapshots created by `zsys`, etc. Funny, I thought I did this earlier, but perhaps I am imagining things.

Thanks for your help :)
